File: php-backend/src/Web/Controller/PlayersDataController.php
<?php


namespace TheScore\Web\Controller;


use DateTimeImmutable;
use GuzzleHttp\Psr7\Response;
use TheScore\PlayerData\Repository\PlayerDataRepositoryInJsonFile;
use TheScore\PlayerData\UseCase\PlayerDataUseCase;
use Throwable;

class PlayersDataController
{

    private const DEFAULT_RECORDS_PER_PAGE = 25;

    public static function getRecordsAsJson(): Response
    {
        try {
            $repo = new PlayerDataRepositoryInJsonFile(__DIR__ . '/../../../data/rushing.json');
            $controller = new PlayerDataUseCase($repo);

            [$nameToFilter, $order, $page, $recordsPerPage] = PlayersDataController::extractParametersFromQueryString();
            $result = $controller->getRecordsAsJson($nameToFilter, $order, $page, $recordsPerPage);
        } catch (Throwable $e) {
            return ErrorController::generateErrorResponse($e);
        }
        return new Response(200, ['Content-Type' => 'application/json'], $result);
    }

    private static function extractParametersFromQueryString(): array
    {
        $params = [];
        parse_str($_SERVER['QUERY_STRING'], $params);

        $nameToFilter = $params['playerName'] ?? '';
        $order = $params['order'] ?? [];
        $page = $params['page'] ?? 1;
        $recordsPerPage = $params['recordsPerPage'] ?? self::DEFAULT_RECORDS_PER_PAGE;
        return [$nameToFilter, $order, $page, $recordsPerPage];
    }

    public static function getRecordsAsDownloadFile(): Response
    {
        try {
            $repo = new PlayerDataRepositoryInJsonFile(__DIR__ . '/../../../data/rushing.json');
            $controller = new PlayerDataUseCase($repo);

            [$nameToFilter, $order, $page, $recordsPerPage] = PlayersDataController::extractParametersFromQueryString();
            $result = $controller->getRecordsAsCsv($nameToFilter, $order, $page, $recordsPerPage);
        } catch (Throwable $e) {
            return ErrorController::generateErrorResponse($e);
        }

        $now = new DateTimeImmutable();
        return new Response(200, [
            'Content-Type' => 'text/csv, application/force-download, application/octet-stream, application/download',
            'Content-Disposition' => "attachment;filename=players_data_{$now->format('U')}.csv",
            'Content-Transfer-Encoding'=> 'binary',
        ], $result);
    }
}

File: elixir-backend/lib/the_rush_web/controllers/players_data_controller.ex
defmodule TheRushWeb.PlayersDataController do
  use TheRushWeb, :controller

  def getPlayers(conn, params) do
    conn = conn
      |> put_resp_header("Access-Control-Allow-Origin", "*")
    render conn, "content_raw.json", response: TheRush.PlayerData.Repository.PlayerDataRepository.getRecords("./data/rushing.json", getPlayerNameParam(params), getPageParam(params), getRecordsPerPageParam(params), getOrderParam(params))
  end

  def exportPlayers(conn, params) do
  conn
  |> put_resp_header("Access-Control-Allow-Origin", "*")
  |> put_resp_content_type("text/csv, application/force-download, application/octet-stream, application/download")
  |> put_resp_header("content-disposition", ~s[attachment; filename="players_data.csv"])
  |> resp(200,TheRush.PlayerData.UseCase.PlayerDataUseCase.getRecordsAsCsv(TheRush.PlayerData.Repository.PlayerDataRepository, "./data/rushing.json", getPlayerNameParam(params), getPageParam(params), getRecordsPerPageParam(params), getOrderParam(params)))
  end


  defp getPageParam(map) when :erlang.is_map_key("page", map) == false, do: 1
  defp getPageParam(%{"page" => page}), do: String.to_integer(page)

  defp getPlayerNameParam(map) when :erlang.is_map_key("playerName", map) == false, do: ""
  defp getPlayerNameParam(%{"playerName" => playerName}), do: playerName

  defp getRecordsPerPageParam(map) when :erlang.is_map_key("recordsPerPage", map) == false, do: 1
  defp getRecordsPerPageParam(%{"recordsPerPage" => recordsPerPage}), do: String.to_integer(recordsPerPage)

  defp getOrderParam(map) when :erlang.is_map_key("order", map) == false, do: []
  defp getOrderParam(%{"order" => order}), do: Enum.map(order, &translateOrderToAtom(&1))

  defp translateOrderToAtom({"Total Rushing Touchdowns", "Desc"}), do: :TotalRushingTouchdownDesc
  defp translateOrderToAtom({"Total Rushing Touchdowns", "Asc"}), do: :TotalRushingTouchdownAsc
  defp translateOrderToAtom({"Longest Rush", "Desc"}), do: :LongestRushDesc
  defp translateOrderToAtom({"Longest Rush", "Asc"}), do: :LongestRushAsc
  defp translateOrderToAtom({"Total Rushing Yards", "Desc"}), do: :TotalRushingYardsDesc
  defp translateOrderToAtom({"Total Rushing Yards", "Asc"}), do: :TotalRushingYardsAsc

end
